man/lsb_release.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.4.
.TH LSB_RELEASE "1" "November 2020" "lsb_release (LSB-Tools-0.9)" "LSB-Tools"
.SH NAME
lsb_release \- manual page for lsb_release (LSB-Tools-0.9)
.SH DESCRIPTION

lsb_release prints certain LSB (Linux Standard Base) and Distribution
infomation.
.PP
usage: lsb_release [\-h] [\-v] [\-i] [\-d] [\-r] [\-c] [\-a] [\-s]
.SS "optional arguments:"
.TP
\fB\-h\fR, \fB\-\-help\fR
show this help message and exit
.TP
\fB\-v\fR, \fB\-\-version\fR
Display the version of the LSB specification against
which the distribution is compliant
.TP
\fB\-i\fR, \fB\-\-id\fR
Display the string id of the distributor
.TP
\fB\-d\fR, \fB\-\-description\fR
Display the single line text description of the
distribution
.TP
\fB\-r\fR, \fB\-\-release\fR
Display the release number of the distribution
.TP
\fB\-c\fR, \fB\-\-codename\fR
Display the codename according to the distribution
release
.TP
\fB\-a\fR, \fB\-\-all\fR
Display all of the above information
.TP
\fB\-s\fR, \fB\-\-short\fR
Display all of the above information in short output
format
.SH FILES

If the installation is LSB compliant, the \fB"/etc/lsb-release"\fR
file should contain the \fBLSB_VERSION\fR field. 
The value of the field should be a colon separated list of
supported module versions indicating the LSB specification 
modules to which the installation is compliant. If the installation is not
compliant, the above field should 
.B not
be present.

Optional fields are \fBDISTRIB_ID\fR, \fBDISTRIB_RELEASE\fR,
\fBDISTRIB_CODENAME\fR, \fBDISTRIB_DESCRIPTION\fR.

If the \fB"/etc/lsb-release.d"\fR directory exists, it is
searched for filenames which are taken as additional module-version
strings to add to \fBLSB_VERSION\fR.
 
The required line style is:
.br 
"\fIDistributor\fR \fBrelease\fR \fIx.x\fR \fB(\fR\fICodename\fR\fB)\fR"

Where \fIDistributor\fR can be a couple of words, but then concatenated
(i.e. Linux Foo BarLinux Linux -> FooBarLinux),
.br
\fIx.x\fR starts with a digit followed by any non-blank characters,
\fICodename\fR will also be concatenated (blanks cleanup only).
.SH EXAMPLES
If the "\fB/etc/lsb-release\fR" file contains:

LSB_VERSION="core-2.0-ia64:core-2.0-noarch"
.br
DISTRIB_DESCRIPTION="I enjoy using my distrib"

and the "\fB/etc/lsb-release.d\fR" directory contains:

graphics-2.0-ia64    graphics-2.0-noarch

Then the results of various options will be:

.B $ ./lsb_release --all
.br
LSB Version:    core-2.0-ia64:core-2.0-noarch:graphics-2.0-ia64:graphics-2.0-noarch
.br
Distributor ID: MyDistrib
.br
Description:    I enjoy using my distrib
.br
Release:        1.0RC4
.br
Codename:       TryIt

.B $ ./lsb_release -a -s
.br
1.0 MyDistrib "I enjoy using my distrib" 1.0RC4 TryIt
.SH AUTHOR
Written by DJ Lucas.
.SH "REPORTING BUGS"
Report bugs at https://github.com/djlucas/LSB-Tools/.
Please include a complete, self contained example that will allow the
bug to be reproduced, and say which version of LSB-Tools you are using.
.SH COPYRIGHT
Copyright \(co 2020 DJ Lucas
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


lsbtools/lsb_release.py
# Begin /usr/bin/lsb_release

import sys
if sys.version_info < (3, 7):
  sys.exit("Python %s.%s or later is required.\n" %(3, 7))

import argparse, glob, itertools, lsbtools, os, re

# Set default values
config = {
  'LSB_VERSION'         : 'unavailable',
  'DISTRIB_ID'          : 'unavailable',
  'DISTRIB_DESCRIPTION' : 'unavailable',
  'DISTRIB_RELEASE'     : 'unavailable',
  'DISTRIB_CODENAME'    : 'unavailable',
}
printval = ''

# Process command line argurments
parser = argparse.ArgumentParser()
parser.add_argument("-v", "--version", help="Display the version of the LSB specification against which the distribution is compliant", action="store_true")
parser.add_argument("-i", "--id", help="Display the string id of the distributor", action="store_true")
parser.add_argument("-d", "--description", help="Display the single line text description of the distribution", action="store_true")
parser.add_argument("-r", "--release", help="Display the release number of the distribution", action="store_true")
parser.add_argument("-c", "--codename", help="Display the codename according to the distribution release", action="store_true")
parser.add_argument("-a", "--all", help="Display all of the above information", action="store_true")
parser.add_argument("-s", "--short", help="Display all of the above information in short output format", action="store_true")
parser.add_argument("--progver", help=argparse.SUPPRESS, action="store_true")
args = parser.parse_args()

if args.progver:
  strver = lsbtools.get_prog_ver(sys.argv[0])
  print(strver, "\n")
  print("Copyright (C) 2020 DJ Lucas")
  print("This is free software; see the source for copying conditions.  There is NO")
  print("warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.")
  print("\nWritten by DJ Lucas.\n")
  sys.exit(0)


if args.version:
  lv = 1
else:
  lv = 0

if args.id:
  li = 1
else:
  li =0

if args.description:
  ld = 1
else:
  ld = 0

if args.release:
  lr = 1
else:
  lr = 0

if args.codename:
  lc = 1
else:
  lc = 0

if args.all:
  lv = 1
  li = 1
  ld = 1
  lr = 1
  lc = 1

if args.short:
  ls = 1
else:
  ls = 0

if lv == 0 and li == 0 and ld == 0 and lr == 0 and lc == 0:
  lv = 1

# Read required configuration file
if not os.path.exists("/etc/lsb-release"):
  print("Required configuration file '/etc/lsb-release' is not found. Exiting...", file=sys.stderr)
  sys.exit(1)

conffile = open("/etc/lsb-release", 'r')
content = conffile.read()
items = content.split('\n')
for pair in items:
  if pair != '':
    key,val = pair.split('=')
    config[key] = val.strip('\"')
conffile.close()

# As of LSB-2.0, the LSB Version string is comprised of colon separated modules
# A module can be represented directly in the LSB_VERSION value or consist
# of empty files with the name of the module in /etc/lsb-release.d/
lsbver = ''
if not os.path.isdir("/etc/lsb-release.d"):
  lsbver = config['LSB_VERSION']
else:
  if len(os.listdir('/etc/lsb-release.d')) == 0:
    lsbver = config['LSB_VERSION']
  else:
    if config['LSB_VERSION'] != 'unavailable':
      lsbver = config['LSB_VERSION']
    # See what else is there
    for lsbfile in os.listdir('/etc/lsb-release.d'):
      if lsbver == '':
        lsbver = basename(lsbfile)
      else:
        lsbver = lsbver + ":" + lsbfile

# Set the LSB Version to our assembled string
config['LSB_VERSION'] = lsbver.strip(' ')

if lv == 1:
  if ls == 1:
    printval = printval + " " + config['LSB_VERSION']
  else:
    lprintval = "LSB Version:\t" + config['LSB_VERSION']
    print(lprintval)

if li == 1:
  if ls == 1:
    printval = printval + " " + config['DISTRIB_ID']
  else:
    lprintval = "Distributor ID:\t" + config['DISTRIB_ID']
    print(lprintval)

if ld == 1:
  if ls == 1:
    printval = printval + " " + config['DISTRIB_DESCRIPTION']
  else:
    lprintval = "Description:\t" + config['DISTRIB_DESCRIPTION']
    print(lprintval)

if lr == 1:
  if ls == 1:
    printval = printval + " " + config['DISTRIB_RELEASE']
  else:
    lprintval = "Release:\t" + config['DISTRIB_RELEASE']
    print(lprintval)

if lc == 1:
  if ls == 1:
    printval = printval + " " + config['DISTRIB_CODENAME']
  else:
    lprintval = "Codename:\t" + config['DISTRIB_CODENAME']
    print(lprintval)

if ls == 1:
  print(printval.strip(" "))

